docs/design-patterns.md

```diff
@@ -0,0 +1,152 @@
+# Design Patterns
+
+The main operating principle for `vdom` is:
+
+> Write functions that return `vdom` elements
+
+In matching with `React` parlance, we'll call these functions components. This allows you to share, remix, and use components amongst everyone. You'll be able to compose together components to create greater layouts than you have
+before.
+
+## Introductory component
+
+We'll start with a component that takes a level of happiness that produces a light visualization:
+
+```python
+from vdom.helpers import div, span, p, meter
+
+def happiness(level=90):
+  smiley = "😃"
+  percent = level / 100.
+    
+  if(percent < 0.25):
+    smiley = "☹️"
+  elif(percent < 0.50):
+    smiley = "😐"
+  elif(percent < 0.75):
+    smiley = "😀"
+
+    
+  return span(
+      p('Happiness ', smiley),
+      meter(level, min=0, low=25, high=75, optimum=90, max=100, value=level)
+  )
+```
+
+The user of the component only has to know how to call `happiness` with their level of happiness:
```

Comment:
The user of the `happiness` component only needs to call it with a level of happiness from 0 to 100.